libbeat/statestore/backend/es/store.go

```diff
@@ -0,0 +1,338 @@
+
+	if err != nil {
+		if status == http.StatusNotFound {
+			return ErrKeyUnknown
+		}
+		return err
+	}
+
+	var qr queryResult
+	err = json.Unmarshal(data, &qr)
+	if err != nil {
+		return err
+	}
+
+	err = json.Unmarshal(qr.Source.Value, to)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+type queryResult struct {
+	Found  bool `json:"found"`
+	Source struct {
+		Value json.RawMessage `json:"v"`
+	} `json:"_source"`
+}
+
+type doc struct {
+	Value     any `struct:"v"`
+	UpdatedAt any `struct:"updated_at"`
+}
+
+type entry struct {
+	value interface{}
+}
+
+func (e entry) Decode(to interface{}) error {
+	return typeconv.Convert(to, e.value)
+}
+
+func renderRequest(val interface{}) doc {
+	return doc{
+		Value:     val,
+		UpdatedAt: time.Now().UTC().Format("2006-01-02T15:04:05.000Z"),
+	}
+}
+
+func (s *store) Set(key string, value interface{}) error {
+	if err := s.waitReady(); err != nil {
+		return err
+	}
+	s.mx.Lock()
+	defer s.mx.Unlock()
+
+	doc := renderRequest(value)
+	_, _, err := s.cli.Request("PUT", fmt.Sprintf("/%s/%s/%s", s.index, docType, url.QueryEscape(key)), "", nil, doc)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *store) Remove(key string) error {
+	if err := s.waitReady(); err != nil {
+		return err
+	}
+	s.mx.Lock()
+	defer s.mx.Unlock()
+
+	_, _, err := s.cli.Delete(s.index, docType, url.QueryEscape(key), nil)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+type searchResult struct {
+	ID     string `json:"_id"`
+	Source struct {
+		Value json.RawMessage `json:"v"`
+	} `json:"_source"`
+}
+
+func (s *store) Each(fn func(string, backend.ValueDecoder) (bool, error)) error {
+	if err := s.waitReady(); err != nil {
+		return err
+	}
+
+	s.mx.Lock()
+	defer s.mx.Unlock()
+
+	// Do nothing for now if the store was not initialized
+	if s.cli == nil {
+		return nil
+	}
+
+	status, result, err := s.cli.SearchURIWithBody(s.index, "", nil, map[string]any{
+		"query": map[string]any{
+			"match_all": map[string]any{},
+		},
+		"size": 1000, // TODO: we might have to do scroll if there are more than 1000 keys
+	})
+
+	if err != nil && status != http.StatusNotFound {
+		return err
+	}
+
+	if result == nil || len(result.Hits.Hits) == 0 {
+		return nil
+	}
+
+	for _, hit := range result.Hits.Hits {
+		var sres searchResult
+		err = json.Unmarshal(hit, &sres)
+		if err != nil {
+			return err
+		}
+
+		var e entry
+		err = json.Unmarshal(sres.Source.Value, &e.value)
+		if err != nil {
+			return err
+		}
+
+		key, err := url.QueryUnescape(sres.ID)
+		if err != nil {
+			return err
+		}
+
+		cont, err := fn(key, e)
+		if !cont || err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s *store) configure(ctx context.Context, c *conf.C) {
+	s.mx.Lock()
+	defer s.mx.Unlock()
+
+	if s.cli != nil {
+		_ = s.cli.Close()
+		s.cli = nil
+	}
+	s.cliErr = nil
+
+	cli, err := eslegclient.NewConnectedClient(ctx, c, s.name)
```

Comment:
Does this mean there is a client and connection per input?

Reply:
Per instance of the storage, yes. Do we want to make it singleton/global?

Reply:
mainly because of this piece:
https://github.com/elastic/beats/blob/09e540c35bcd27e5962a98fc96acd2d431d5424a/libbeat/esleg/eslegclient/connection.go#L253
```
// NewConnectedClient returns a non-thread-safe connection. Make sure for each goroutine you initialize a new connection.
func NewConnectedClient(ctx context.Context, cfg *cfg.C, beatname string) (*Connection, error) {
```